source/SIL.AppBuilder.Portal.Frontend/src/ui/routes/projects/list/header/bulk-product-selection/product-selection.tsx

```diff
@@ -0,0 +1,177 @@
+                    >
+                      <Checkbox
+                        data-test-item-checkbox
+                        className='m-r-sm'
+                        value={id}
+                        checked={isProductSelected}
+                      />
+                      <ProductIcon
+                        product={productDefinition}
+                        selected={isProductSelected}
+                        size={19}
+                      />
+                      <span
+                        data-test-item-text
+                        className={`p-l-xs-xs ${isSelected && 'black-text'}`}
+                      >
+                        {name}
+                      </span>
+                    </div>
+                    <div className='w-100 p-sm p-t-md p-b-md fs-11'>
+                      <span>{description}</span>
+                    </div>
+                  </div>
+                );
+              })}
+            </>
+          );
+        }}
+      </AsyncWaiter>
+    </ErrorBoundary>
+  );
+}
+
+interface IAllowedBulkList {
+  productDefinition: ProductDefinitionResource;
+  isAllowed: boolean;
+}
+
+function useProductSelection(
+  selectedRows: ProjectResource[],
+  dataStore: Store,
+  selected: any,
+  onChange: (ids: string[]) => void
+) {
+  useEffect(() => {
+    // given the selected projects, find the products that also have any of the
+    // selected product definitions
+    let products = [];
+
+    selectedRows.forEach((project) => {
+      const projectProducts = dataStore.cache
+        .query((q) => q.findRelatedRecords(project, 'products'))
+        .filter((product) => {
+          const productDefinition = dataStore.cache.query((q) =>
+            q.findRelatedRecord(product, 'productDefinition')
+          );
+          // this comparison is using local ids-
+          return selected.includes(productDefinition.id);
+        });
+
+      products = products.concat(projectProducts);
+    });
+
+    // remote ids
+    const productIds = products.map((product) => idFromRecordIdentity(dataStore, product));
+
+    onChange(productIds);
+  }, [dataStore, onChange, selected, selectedRows]);
+}
+
+function uesPermissions(
+  selectedRows: ProjectResource[],
+  dataStore: Store,
+  onPermissionRetrieval: (newPermissions: Permissions) => void
+) {
+  return useCallback(async () => {
+    const ids = selectedRows.map((project) => idFromRecordIdentity(dataStore, project));
+    const response = await authenticatedPost(`/api/product-actions`, {
+      data: { projects: ids },
+      headers: {
+        ['Content-Type']: 'application/json',
+      },
+    });
+
+    if (response.status >= 400) {
+      const json = await tryParseJson(response);
+      throw new GenericJsonApiError(response.status, response.statusText, json);
+    }
+
+    const json = await response.json();
+
+    onPermissionRetrieval(json);
+
+    return json;
+  }, [dataStore, onPermissionRetrieval, selectedRows]);
+}
```

Comment:
apologies again for the long file. on the plus side, all these functions are kinda one-time use -- mostly for breaking up the main component